File: android/src/main/java/com/resideo/auth0_flutter/AuthenticationController.java
package com.resideo.auth0_flutter;

import androidx.annotation.NonNull;

import com.auth0.android.Auth0;
import com.auth0.android.authentication.AuthenticationAPIClient;
import com.auth0.android.authentication.AuthenticationException;
import com.auth0.android.authentication.PasswordlessType;
import com.auth0.android.authentication.request.DatabaseConnectionRequest;
import com.auth0.android.callback.AuthenticationCallback;
import com.auth0.android.callback.BaseCallback;
import com.auth0.android.request.AuthenticationRequest;
import com.auth0.android.result.Credentials;
import com.auth0.android.result.DatabaseUser;
import com.auth0.android.result.UserProfile;

import java.util.HashMap;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

public class AuthenticationController implements MethodCallHandler {
    private final Registrar registrar;

    static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "plugins.auth0_flutter.io/authentication");

        channel.setMethodCallHandler(new AuthenticationController(registrar));
    }

    private AuthenticationController(Registrar registrar) { this.registrar = registrar; }

    @Override
    public void onMethodCall(MethodCall call, @NonNull final Result result) {
        final String clientId = call.argument("clientId");
        final String domain = call.argument("domain");

        assert clientId != null;
        assert domain != null;

        final Auth0 auth0 = new Auth0(clientId, domain);
        auth0.setOIDCConformant(true);

        AuthenticationAPIClient authClient = new AuthenticationAPIClient(auth0);

        switch (call.method) {
            case AuthMethodName.login: {
                final String usernameOrEmail = call.argument("usernameOrEmail");
                final String password = call.argument("password");
                final String realm = call.argument("realm");

                assert usernameOrEmail != null;
                assert password != null;
                assert realm != null;

                final String audience = call.argument("audience");
                final String scope = call.argument("scope");

                final AuthenticationRequest request = authClient.login(usernameOrEmail, password, realm);

                if (audience != null) {
                    request.setAudience(audience);
                }

                if (scope != null) {
                    request.setScope(scope);
                }

                HashMap<String, Object> params = call.argument("parameters");
                if (params != null) {
                    request.addAuthenticationParameters(params);
                }

                request.start(new BaseCallback<Credentials, AuthenticationException>() {
                    @Override
                    public void onSuccess(Credentials payload) {
                        final HashMap<String, Object> obj = JSONHelpers.credentialsToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.loginWithOTP: {
                final String otp = call.argument("otp");
                final String mfaToken = call.argument("mfaToken");

                assert otp != null;
                assert mfaToken != null;

                authClient.loginWithOTP(mfaToken, otp).start(new BaseCallback<Credentials, AuthenticationException>() {
                    @Override
                    public void onSuccess(Credentials payload) {
                        final HashMap<String, Object> obj = JSONHelpers.credentialsToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }
//            case AuthMethodName.loginDefaultDirectory: {
//                result.notImplemented();
//                break;
//            }

            case AuthMethodName.createUser: {
                final String email = call.argument("email");
                final String username = call.argument("username");
                final String password = call.argument("password");
                final String connection = call.argument("connection");

                assert email != null;
                assert password != null;
                assert connection != null;

                final DatabaseConnectionRequest<DatabaseUser, AuthenticationException> request;

                if (username != null) {
                    request = authClient.createUser(email, password, username, connection);
                } else {
                    request = authClient.createUser(email, password, connection);
                }

                request.start(new BaseCallback<DatabaseUser, AuthenticationException>() {
                    @Override
                    public void onSuccess(DatabaseUser payload) {
                        final HashMap<String, Object> obj = JSONHelpers.databaseUserToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.resetPassword: {
                final String email = call.argument("email");
                final String connection = call.argument("connection");

                assert email != null;
                assert connection != null;

                authClient.resetPassword(email, connection).start(new BaseCallback<Void, AuthenticationException>() {
                    @Override
                    public void onSuccess(Void payload) {
                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(null);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.startEmailPasswordless: {
                final String email = call.argument("email");
                final String typeString = call.argument("type");

                assert email != null;
                assert typeString != null;

                final PasswordlessType type = PasswordlessType.valueOf(typeString);

                authClient.passwordlessWithEmail(email, type).start(new BaseCallback<Void, AuthenticationException>() {
                    @Override
                    public void onSuccess(Void payload) {
                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(null);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.startPhoneNumberPasswordless: {
                final String phoneNumber = call.argument("phoneNumber");
                final String typeString = call.argument("type");

                assert phoneNumber != null;
                assert typeString != null;

                final PasswordlessType type = PasswordlessType.valueOf(typeString);

                authClient.passwordlessWithSMS(phoneNumber, type).start(new BaseCallback<Void, AuthenticationException>() {
                    @Override
                    public void onSuccess(Void payload) {
                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(null);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.loginEmailPasswordless: {
                final String email = call.argument("email");
                final String code = call.argument("code");

                assert email != null;
                assert code != null;

                authClient.loginWithEmail(email, code).start(new BaseCallback<Credentials, AuthenticationException>() {
                    @Override
                    public void onSuccess(Credentials payload) {
                        final HashMap<String, Object> obj = JSONHelpers.credentialsToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.loginPhoneNumberPasswordless: {
                final String phoneNumber = call.argument("phoneNumber");
                final String code = call.argument("code");

                assert phoneNumber != null;
                assert code != null;

                authClient.loginWithPhoneNumber(phoneNumber, code).start(new BaseCallback<Credentials, AuthenticationException>() {
                    @Override
                    public void onSuccess(Credentials payload) {
                        final HashMap<String, Object> obj = JSONHelpers.credentialsToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.userInfoWithToken: {
                final String token = call.argument("token");

                assert token != null;

                authClient.tokenInfo(token).start(new BaseCallback<UserProfile, AuthenticationException>() {
                    @Override
                    public void onSuccess(UserProfile payload) {
                        final HashMap<String, Object> obj = JSONHelpers.profileTOJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.userInfoWithAccessToken: {
                final String accessToken = call.argument("accessToken");
                assert accessToken != null;

                authClient.userInfo(accessToken).start(new AuthenticationCallback<UserProfile>() {
                    @Override
                    public void onSuccess(UserProfile payload) {
                        final HashMap<String, Object> obj = JSONHelpers.profileTOJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.loginSocial: {
                final String token = call.argument("token");
                final String connection = call.argument("connection");

                assert token != null;
                assert connection != null;

                final AuthenticationRequest request = authClient.loginWithOAuthAccessToken(token, connection);

                request.start(new BaseCallback<Credentials, AuthenticationException>() {
                    @Override
                    public void onSuccess(Credentials payload) {
                        final HashMap<String, Object> obj = JSONHelpers.credentialsToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }
//            case AuthMethodName.tokenExchangeWithParameters: {
//                result.notImplemented();
//                break;
//            }
//            case AuthMethodName.tokenExchangeWithCode: {
//                result.notImplemented();
//                break;
//            }
//            case AuthMethodName.appleTokenExchange: {
//                result.notImplemented();
//                break;
//            }

            case AuthMethodName.renew: {
                final String refreshToken = call.argument("refreshToken");

                assert refreshToken != null;

                authClient.renewAuth(refreshToken).start(new AuthenticationCallback<Credentials>() {
                    @Override
                    public void onSuccess(Credentials payload) {
                        final HashMap<String, Object> obj = JSONHelpers.credentialsToJSON(payload);

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                result.success(obj);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() { result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }

            case AuthMethodName.revoke: {
                final String refreshToken = call.argument("refreshToken");

                assert refreshToken != null;

                authClient.revokeToken(refreshToken).start(new AuthenticationCallback<Void>() {
                    @Override
                    public void onSuccess(Void payload) {
                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() { result.success(true);
                            }
                        });
                    }

                    @Override
                    public void onFailure(AuthenticationException error) {
                        final HashMap<String, Object> obj = JSONHelpers.authErrorToJSON(error);

                        final String message = error.getLocalizedMessage();

                        registrar.activity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() { result.error("", message, obj);
                            }
                        });
                    }
                });
                break;
            }
//            case AuthMethodName.delegation: {
//                result.notImplemented();
//                break;
//            }
            default:
                result.notImplemented();
                break;
        }
    }
}

class AuthMethodName {
    static final String login = "login";
    static final String loginWithOTP = "loginWithOtp";
    static final String loginDefaultDirectory = "loginDefaultDirectory";
    static final String createUser = "createUser";
    static final String resetPassword = "resetPassword";
    static final String startEmailPasswordless = "startEmailPasswordless";
    static final String startPhoneNumberPasswordless = "startPhoneNumberPasswordless";
    static final String loginEmailPasswordless = "loginEmailPasswordless";
    static final String loginPhoneNumberPasswordless = "loginPhoneNumberPasswordless";
    static final String userInfoWithToken = "userInfoWithToken";
    static final String userInfoWithAccessToken = "userInfoWithAccessToken";
    static final String loginSocial = "loginSocial";
    static final String tokenExchangeWithParameters = "tokenExchangeWithParams";
    static final String tokenExchangeWithCode = "tokenExchangeWithCode";
    static final String appleTokenExchange = "appleTokenExchange";
    static final String renew = "renew";
    static final String revoke = "revoke";
    static final String delegation = "delegation";
}

File: ios/Classes/Controllers/AuthenticationController.swift
//
//  AuthenticationController.swift
//  auth0_flutter
//
//  Created by Juan Alvarez on 9/24/19.
//

import Foundation
import Auth0

class AuthenticationController: NSObject, FlutterPlugin {
    enum MethodName: String {
        case login
        case loginWithOTP
        case loginDefaultDirectory
        case createUser
        case resetPassword
        case startEmailPasswordless
        case startPhoneNumberPasswordless
        case userInfoWithToken
        case userInfoWithAccessToken
        case loginSocial
        case tokenExchangeWithParameters
        case tokenExchangeWithCode
        case appleTokenExchange
        case renew
        case revoke
        case delegation
    }
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "plugins.auth0_flutter.io/authentication", binaryMessenger: registrar.messenger())
        
        let instance = AuthenticationController()
        
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments ?? [:]
        
        do {
            guard let method = MethodName(rawValue: call.method) else {
                sendResult(result, data: nil, error: Auth0PluginError.unknownMethod(call.method))
                return
            }
            
            let data = try JSONSerialization.data(withJSONObject: arguments, options: [])
            let authParams = try AuthParameters.decode(data: data)
            
            let auth = authentication(clientId: authParams.clientId, domain: authParams.domain)
            
            switch method {
            case .login:
                let params = try LoginWithUsernameOrEmailParameters.decode(data: data)
                auth.login(usernameOrEmail: params.usernameOrEmail,
                           password: params.password,
                           realm: params.realm,
                           audience: params.audience,
                           scope: params.scope,
                           parameters: params.parametersDict())
                    .start { (authResult) in
                        switch authResult {
                        case .success(let credentials):
                            sendResult(result, data: credentials.toJSON(), error: nil)
                        case .failure(let error):
                            sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                        }
                }
                
            case .loginWithOTP:
                let params = try LoginWithOTPParameters.decode(data: data)
                auth.login(withOTP: params.otp, mfaToken: params.mfaToken).start { (authResult) in
                    switch authResult {
                    case .success(let credentials):
                        sendResult(result, data: credentials.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .loginDefaultDirectory:
                let params = try LoginWithDefaultDirectoryParameters.decode(data: data)
                auth.loginDefaultDirectory(withUsername: params.username,
                                           password: params.password,
                                           audience: params.audience,
                                           scope: params.scope,
                                           parameters: params.parametersDict())
                    .start { (authResult) in
                        switch authResult {
                        case .success(let credentials):
                            sendResult(result, data: credentials.toJSON(), error: nil)
                        case .failure(let error):
                            sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                        }
                }
                
            case .createUser:
                let params = try CreateUserParameters.decode(data: data)
                auth.createUser(email: params.email,
                                username: params.username,
                                password: params.password,
                                connection: params.connection,
                                userMetadata: params.userMetadataDict(),
                                rootAttributes: params.rootAttributesDict())
                    .start { (authResult) in
                        switch authResult {
                        case .success(let user):
                            sendResult(result, data: databaseUserToJSON(user), error: nil)
                        case .failure(let error):
                            sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                        }
                }
                
            case .resetPassword:
                let params = try ResetPasswordParameters.decode(data: data)
                auth.resetPassword(email: params.email, connection: params.connection).start { (authResult) in
                    switch authResult {
                    case .success(_):
                        sendResult(result, data: nil, error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .startEmailPasswordless:
                let params = try StartEmailPasswordlessParameters.decode(data: data)
                auth.startPasswordless(email: params.email, type: params.type, connection: params.connection, parameters: params.parametersDict()).start { (authResult) in
                    switch authResult {
                    case .success(_):
                        sendResult(result, data: nil, error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .startPhoneNumberPasswordless:
                let params = try StartPhoneNumberPasswordlessParameters.decode(data: data)
                auth.startPasswordless(phoneNumber: params.phoneNumber, type: params.type, connection: params.connection).start { (authResult) in
                    switch authResult {
                    case .success(_):
                        sendResult(result, data: nil, error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .userInfoWithToken:
                let params = try UserInfoWithTokenParameters.decode(data: data)
                auth.userInfo(token: params.token).start { (authResult) in
                    switch authResult {
                    case .success(let profile):
                        sendResult(result, data: profile.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .userInfoWithAccessToken:
                let params = try UserInfoWithAccessTokenParameters.decode(data: data)
                auth.userInfo(withAccessToken: params.accessToken).start { (authResult) in
                    switch authResult {
                    case .success(let userInfo):
                        sendResult(result, data: userInfo.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .loginSocial:
                let params = try LoginSocialParameters.decode(data: data)
                auth.loginSocial(token: params.token, connection: params.connection, scope: params.scope, parameters: params.parametersDict()).start { (authResult) in
                    switch authResult {
                    case .success(let credentials):
                        sendResult(result, data: credentials.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .tokenExchangeWithParameters:
                let params = try TokenExchangeWithParamsParameters.decode(data: data)
                auth.tokenExchange(withParameters: params.parametersDict()).start { (authResult) in
                    switch authResult {
                    case .success(let credentials):
                        sendResult(result, data: credentials.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .tokenExchangeWithCode:
                let params = try TokenExchangeWithCodeParameters.decode(data: data)
                auth.tokenExchange(withCode: params.code, codeVerifier: params.code, redirectURI: params.redirectURI).start { (authResult) in
                    switch authResult {
                    case .success(let credentials):
                        sendResult(result, data: credentials.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .appleTokenExchange:
                let params = try AppleTokenExchangeParameters.decode(data: data)
                auth.tokenExchange(withAppleAuthorizationCode: params.authCode, scope: params.scope, audience: params.audience).start { (authResult) in
                    switch authResult {
                    case .success(let credentials):
                        sendResult(result, data: credentials.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .renew:
                let params = try RenewParameters.decode(data: data)
                auth.renew(withRefreshToken: params.refreshToken, scope: params.scope).start { (authResult) in
                    switch authResult {
                    case .success(let credentials):
                        sendResult(result, data: credentials.toJSON(), error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .revoke:
                let params = try RevokeParameters.decode(data: data)
                auth.revoke(refreshToken: params.refreshToken).start { (authResult) in
                    switch authResult {
                    case .success(_):
                        sendResult(result, data: nil, error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
                
            case .delegation:
                let params = try DelegationParameters.decode(data: data)
                auth.delegation(withParameters: params.parametersDict()).start { (authResult) in
                    switch authResult {
                    case .success(let dict):
                        sendResult(result, data: dict, error: nil)
                    case .failure(let error):
                        sendResult(result, data: nil, error: CustomAuthenticationError(error as! AuthenticationError))
                    }
                }
            }
        } catch {
            sendResult(result, data: nil, error: error)
        }
    }
}

private struct AuthParameters: Decodable {
    let clientId: String
    let domain: String
}

private struct LoginWithUsernameOrEmailParameters: Decodable {
    let usernameOrEmail: String
    let password: String
    let realm: String
    let audience: String?
    let scope: String?
    private let parameters: [String: AnyDecodable]?
    
    func parametersDict() -> [String: Any]? {
        return parameters?.mapValues { $0.value }
    }
}

private struct LoginWithOTPParameters: Decodable {
    let otp: String
    let mfaToken: String
}

private struct LoginWithDefaultDirectoryParameters: Decodable {
    let username: String
    let password: String
    let audience: String?
    let scope: String?
    private let parameters: [String: AnyDecodable]?
    
    func parametersDict() -> [String: Any]? {
        return parameters?.mapValues { $0.value }
    }
}

private struct CreateUserParameters: Decodable {
    let email: String
    let username: String?
    let password: String
    let connection: String
    private let userMetadata: [String: AnyDecodable]?
    private let rootAttributes: [String: AnyDecodable]?
    
    func userMetadataDict() -> [String: Any]? {
        return userMetadata?.mapValues { $0.value }
    }
    
    func rootAttributesDict() -> [String: Any]? {
        return rootAttributes?.mapValues { $0.value }
    }
}

private struct ResetPasswordParameters: Decodable {
    let email: String
    let connection: String
}

extension PasswordlessType: Decodable {}

private struct StartEmailPasswordlessParameters: Decodable {
    let email: String
    let type: PasswordlessType
    let connection: String
    private let parameters: [String: AnyDecodable]
    
    func parametersDict() -> [String: Any] {
        return parameters.mapValues { $0.value }
    }
}

private struct StartPhoneNumberPasswordlessParameters: Decodable {
    let phoneNumber: String
    let type: PasswordlessType
    let connection: String
}

private struct UserInfoWithTokenParameters: Decodable {
    let token: String
}

private struct UserInfoWithAccessTokenParameters: Decodable {
    let accessToken: String
}

private struct LoginSocialParameters: Decodable {
    let token: String
    let connection: String
    let scope: String
    private let parameters: [String: AnyDecodable]
    
    func parametersDict() -> [String: Any] {
        return parameters.mapValues { $0.value }
    }
}

private struct TokenExchangeWithParamsParameters: Decodable {
    private let parameters: [String: AnyDecodable]
    
    func parametersDict() -> [String: Any] {
        return parameters.mapValues { $0.value }
    }
}

private struct TokenExchangeWithCodeParameters: Decodable {
    let code: String
    let codeVerifier: String
    let redirectURI: String
}

private struct AppleTokenExchangeParameters: Decodable {
    let authCode: String
    let scope: String?
    let audience: String?
}

private struct RenewParameters: Decodable {
    let refreshToken: String
    let scope: String?
}

private struct RevokeParameters: Decodable {
    let refreshToken: String
}

private struct DelegationParameters: Decodable {
    private let parameters: [String: AnyDecodable]
    
    func parametersDict() -> [String: Any] {
        return parameters.mapValues { $0.value }
    }
}
